core/controllers/skill_editor_test.py

```diff
@@ -52,41 +79,75 @@ def test_access_skill_editor_page(self):
+            response = self.testapp.get(self.url)
             self.assertEqual(response.status_int, 200)
             self.logout()
 
-    def test_editable_skill_handler_get(self):
-        # Check that non-admins cannot access the editable skill data.
-        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
-            self.login(self.NEW_USER_EMAIL)
-            response = self.testapp.get(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_DATA_URL_PREFIX, self.skill_id),
-                expect_errors=True)
-            self.assertEqual(response.status_int, 401)
-            self.logout()
+    def test_skill_editor_page_fails(self):
+        self.login(self.ADMIN_EMAIL)
+        # Check GET returns 404 when new strutures' pages are not enabled.
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', False):
+            response = self.testapp.get(self.url, expect_errors=True)
+            self.assertEqual(response.status_int, 404)
 
-            # Check that admins can access the editable skill data.
+        # Check GET returns 404 when cannot get skill by id.
+        feconf_swap = self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True)
+        skill_services_swap = self.swap(
+            skill_services, 'get_skill_by_id', self._mock_get_skill_by_id)
+        with feconf_swap, skill_services_swap:
             self.login(self.ADMIN_EMAIL)
-            json_response = self.get_json(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_DATA_URL_PREFIX, self.skill_id))
-            self.assertEqual(self.skill_id, json_response['skill']['id'])
-            self.logout()
+            response = self.testapp.get(self.url, expect_errors=True)
+            self.assertEqual(response.status_int, 404)
+        self.logout()
+
 
-    def test_editable_skill_handler_put(self):
-        # Check that admins can edit a skill.
-        change_cmd = {
+class SkillRightsHandlerTest(BaseSkillEditorControllerTest):
+    """Tests for SkillRightsHandler."""
+
+    def setUp(self):
+        super(SkillRightsHandlerTest, self).setUp()
+        self.url = '%s/%s' % (feconf.SKILL_RIGHTS_URL_PREFIX, self.skill_id)
+
+    def test_skill_rights_handler_succeeds(self):
+        self.login(self.ADMIN_EMAIL)
+        # Check that admins can access and edit in the editor page.
+        self.get_json(self.url)
+        # Check GET returns JSON object with can_edit_skill_description set
+        # to False if the user is not allowed to edit the skill description.
+        def _get_all_actions(*_args):
+            actions = list(self.admin.actions)
+            actions.remove(role_services.ACTION_EDIT_SKILL_DESCRIPTION)
+            return actions
+        with self.swap(role_services, 'get_all_actions', _get_all_actions):
+            json_resp = self.get_json(self.url)
+            self.assertEqual(json_resp['can_edit_skill_description'], False)
+        self.logout()
+
+    def test_skill_rights_handler_fails(self):
```

Comment:
@seanlip This test aims to cover this line https://github.com/oppia/oppia/blob/a325d544ddfa240a43c7ca3046111164bfd2402b/core/controllers/skill_editor.py#L104

I've spent quite a lot of time making the test cover that lines without success. Perhaps you can figure out what is missing in the setup?

Reply:
@aks681 Can you help me with this issue?

Reply:
I am not sure why that does not work, but can you try @seanlip's comments below about using an invalid skill_id directly and check if that works?
Also, to try troubleshooting, see the docstring for the swap function itself here:
https://github.com/oppia/oppia/blob/a325d544ddfa240a43c7ca3046111164bfd2402b/core/tests/test_utils.py#L1100

Maybe that will help.

Reply:
@seanlip @aks681 
I want to follow up with Sean's suggestion for this test.

I tried this
```
model = skill_models.SkillRightsModel.get(self.skill_id)
model.delete(self.admin_id, 'Deleting the model.')
```

before making the GET request. I also put debugging lines in `skill_services.get_skill_rights` and it did return `None`. However, the debugging line on `skill_editor` was not printed out on the screen and the line is still not covered.

One thing worth notice is the return error is still as expected (404). If the line of code that we're concerned with is not reached, where in the code that an exception was raised that lead to the 404 error?

Reply:
I think there are two things going on here.

Firstly, your code might be hitting the decorator (see the comment I just left previously) since you haven't created a skill.

Secondly, the suggestion I have is actually to create a skill the normal way (using skill_services.save_new_skill) and then manually delete the rights model. I think that should (?) yield the desired result.

Reply:
@seanlip I think this is dead code because the decorator code checks the same thing here

https://github.com/oppia/oppia/blob/f3cf204eeda079ee284936e7c8b8bee56bb56d15/core/domain/acl_decorators.py#L1417

Even if we follow your second suggestion, those lines of code in `skill_editor.py` would not be reached anyway.

Let's delete the dead code?

Reply:
Done